### PR TITLE
Added support for storpool_qos service 

### PR DESCRIPTION
**Pull request by slavkap**
### Description

This PR provides support to use the storpool_qos service (to change tiers) via the Disk offering's details.
StorPool provides the ‘storpool_qos’ service that tracks and configures the storage tier for all volumes based on a specifically provided `qc` tag specifying the storage tier for each volume.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

### How Has This Been Tested?
manual and smoke tests

### CODE REVIEW

File: plugins/storage/volume/storpool/src/main/java/org/apache/cloudstack/storage/datastore/driver/StorPoolPrimaryDataStoreDriver.java [hunk 1]
```diff
@@ -328,66 +373,107 @@ private StorPoolSetVolumeEncryptionAnswer createEncryptedVolume(DataStore dataSt
...
-                err = e.toString();
             }
+        } catch (Exception e) {
+            logger.debug("sending resize command failed", e);
+            err = e.toString();
+        }
+        return err;
+    }
+
+    private String notifyQemuForTheNewSize(DataObject data, String err, VolumeObject vol, ResizeVolumePayload payload)
```
Comment (sureshanaparti) [inline plugins/storage/volume/storpool/src/main/java/org/apache/cloudstack/storage/datastore/driver/StorPoolPrimaryDataStoreDriver.java:421]:
```suggestion
    private String notifyQemuForTheNewSize(DataObject data, VolumeObject vol, ResizeVolumePayload payload)
```

seems `err` not needed in the call

File: plugins/storage/volume/storpool/src/main/java/org/apache/cloudstack/storage/datastore/driver/StorPoolPrimaryDataStoreDriver.java [hunk 2]
```diff
@@ -328,66 +373,107 @@ private StorPoolSetVolumeEncryptionAnswer createEncryptedVolume(DataStore dataSt
...
+                payload.instanceName, payload.hosts == null ? false : true);
+        ResizeVolumeAnswer answer = (ResizeVolumeAnswer) storageMgr.sendToPool(pool, payload.hosts, resizeCmd);
+
+        if (answer == null || !answer.getResult()) {
+            err = answer != null ? answer.getDetails() : "return a null answer, resize failed for unknown reason";
+        }
+        return err;
+    }
+
+    private String updateStorPoolVolume(String err, VolumeObject vol, ResizeVolumePayload payload, SpConnectionDesc conn) {
```
Comment (sureshanaparti) [inline plugins/storage/volume/storpool/src/main/java/org/apache/cloudstack/storage/datastore/driver/StorPoolPrimaryDataStoreDriver.java:435]:
```suggestion
    private String updateStorPoolVolume(VolumeObject vol, ResizeVolumePayload payload, SpConnectionDesc conn) {
```

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 3]
```diff
@@ -2324,7 +2339,7 @@ private void validateVolumeResizeWithSize(VolumeVO volume, long currentSize, Lon
              * the actual disk size.
              */
             if (currentSize > newSize) {
-                if (volume != null && ImageFormat.QCOW2.equals(volume.getFormat()) && !Volume.State.Allocated.equals(volume.getState())) {
+                if (volume != null && ImageFormat.QCOW2.equals(volume.getFormat()) && !Volume.State.Allocated.equals(volume.getState()) && !StoragePoolType.StorPool.equals(volume.getPoolType())) {
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:2342]: hard-coded check, can check this through StorPool storage plugin?

Comment (slavkap) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:2342]: @sureshanaparti, there are a lot similar checks in the code 
https://github.com/apache/cloudstack/blob/6dc3d06037c39019f29686281856443c37a3e6c0/server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java#L3229

I think there is no need to complicate the code by adding additional method in the storage drivers

Comment (sureshanaparti) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:2342]: agree @slavkap , maybe these hard-coded checks have to be fixed at some point of time with proper testing.

**Review by sureshanaparti — APPROVED** [review]
clgtm

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@rp- @rg9975 @harikrishna-patnala can you review and maybe this? there is albeit minor changes in the generic code.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

File: test/integration/plugins/storpool/TestStorPoolTiers.py
Comment (DaanHoogland) [inline test/integration/plugins/storpool/TestStorPoolTiers.py:1]: should the file name adhere to python conventions more (no camelcase but underscores)?

File: test/integration/plugins/storpool/TestStorPoolTiers.py [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+        cls.cleanUpCloudStack()
+
+    @classmethod
+    def cleanUpCloudStack(cls):
+        try:
+            # Cleanup resources used
+            cleanup_resources(cls.apiclient, cls._cleanup)
+        except Exception as e:
+            raise Exception("Warning: Exception during cleanup : %s" % e)
+        return
```
Comment (DaanHoogland) [inline test/integration/plugins/storpool/TestStorPoolTiers.py:195]:
```suggestion
        super(TestStorPoolTiers, cls).tearDownClass()
```

File: test/integration/plugins/storpool/TestStorPoolTiers.py [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+
+    def setUp(self):
+        self.apiclient = self.testClient.getApiClient()
+        self.dbclient = self.testClient.getDbConnection()
+
+        if self.unsupportedHypervisor:
+            self.skipTest("Skipping test because unsupported hypervisor\
+                    %s" % self.hypervisor)
+        return
```
Comment (DaanHoogland) [inline test/integration/plugins/storpool/TestStorPoolTiers.py:204]:
```suggestion
        self.cleanup = []
        return
```

File: test/integration/plugins/storpool/TestStorPoolTiers.py [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+
+    def tearDown(self):
+        return
```
Comment (DaanHoogland) [inline test/integration/plugins/storpool/TestStorPoolTiers.py:207]:
```suggestion
        super(TestStorPoolTiers, self).tearDown()
        return
```

File: test/integration/plugins/storpool/TestStorPoolTiers.py [hunk 7]
```diff
@@ -0,0 +1,554 @@
...
+            templateid=self.template.id,
+            accountid=self.account.name,
+            domainid=self.account.domainid,
+            serviceofferingid=self.service_offering.id,
+            overridediskofferingid=self.disk_offerings_tier1_template.id,
+            diskofferingid=self.disk_offerings_tier1_template.id,
+            size=2,
+            hypervisor=self.hypervisor,
+            rootdisksize=10
+        )
```
Comment (DaanHoogland) [inline test/integration/plugins/storpool/TestStorPoolTiers.py:254]:
```suggestion
        virtual_machine_template_tier1 = VirtualMachine.create(
            self.apiclient,
            {"name": "StorPool-%s" % uuid.uuid4()},
            zoneid=self.zone.id,
            templateid=self.template.id,
            accountid=self.account.name,
            domainid=self.account.domainid,
            serviceofferingid=self.service_offering.id,
            overridediskofferingid=self.disk_offerings_tier1_template.id,
            diskofferingid=self.disk_offerings_tier1_template.id,
            size=2,
            hypervisor=self.hypervisor,
            rootdisksize=10
        )
        self.cleanup.append(virtual_machine_template_tier1)
```

File: test/integration/plugins/storpool/TestStorPoolTiers.py [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+            templateid=self.template.id,
+            accountid=self.account.name,
+            domainid=self.account.domainid,
+            serviceofferingid=self.service_offering.id,
+            overridediskofferingid=self.disk_offerings_tier2_template.id,
+            diskofferingid=self.disk_offerings_tier2_template.id,
+            size=2,
+            hypervisor=self.hypervisor,
+            rootdisksize=10
+        )
```
Comment (DaanHoogland) [inline test/integration/plugins/storpool/TestStorPoolTiers.py:275]:
```suggestion
        virtual_machine_template_tier2 = VirtualMachine.create(
            self.apiclient,
            {"name": "StorPool-%s" % uuid.uuid4()},
            zoneid=self.zone.id,
            templateid=self.template.id,
            accountid=self.account.name,
            domainid=self.account.domainid,
            serviceofferingid=self.service_offering.id,
            overridediskofferingid=self.disk_offerings_tier2_template.id,
            diskofferingid=self.disk_offerings_tier2_template.id,
            size=2,
            hypervisor=self.hypervisor,
            rootdisksize=10
        )
        self.cleanup.append(virtual_machine_template_tier2)
```

File: test/integration/plugins/storpool/TestStorPoolTiers.py [hunk 9]
```diff
@@ -0,0 +1,554 @@
...
+            templateid=self.template.id,
+            accountid=self.account.name,
+            domainid=self.account.domainid,
+            serviceofferingid=self.service_offering.id,
+            overridediskofferingid=disk_off_id,
+            diskofferingid=disk_off_id,
+            size=2,
+            hypervisor=self.hypervisor,
+            rootdisksize=10
+        )
```
Comment (DaanHoogland) [inline test/integration/plugins/storpool/TestStorPoolTiers.py:297]:
```suggestion
        virtual_machine_tier2_tag = VirtualMachine.create(
            self.apiclient,
            {"name": "StorPool-%s" % uuid.uuid4()},
            zoneid=self.zone.id,
            templateid=self.template.id,
            accountid=self.account.name,
            domainid=self.account.domainid,
            serviceofferingid=self.service_offering.id,
            overridediskofferingid=disk_off_id,
            diskofferingid=disk_off_id,
            size=2,
            hypervisor=self.hypervisor,
            rootdisksize=10
        )
        self.cleanup.append(virtual_machine_tier2_tag)
```

File: test/integration/plugins/storpool/TestStorPoolTiers.py [hunk 10]
```diff
@@ -0,0 +1,554 @@
...
```
Comment (DaanHoogland) [inline test/integration/plugins/storpool/TestStorPoolTiers.py:330]:
```suggestion
        virtual_machine_tier2_template = VirtualMachine.create(
            self.apiclient,
            {"name": "StorPool-%s" % uuid.uuid4()},
            zoneid=self.zone.id,
            templateid=self.template.id,
            accountid=self.account.name,
            domainid=self.account.domainid,
            serviceofferingid=self.service_offering.id,
            overridediskofferingid=disk_off_id,
            diskofferingid=disk_off_id,
            size=2,
            hypervisor=self.hypervisor,
            rootdisksize=10
        )
        self.cleanup.append(virtual_machine_tier2_template)
```

File: test/integration/plugins/storpool/TestStorPoolTiers.py [hunk 11]
```diff
@@ -0,0 +1,554 @@
...
```
Comment (DaanHoogland) [inline test/integration/plugins/storpool/TestStorPoolTiers.py:365]:
```suggestion
        virtual_machine_tier2_template = VirtualMachine.create(
            self.apiclient,
            {"name": "StorPool-%s" % uuid.uuid4()},
            zoneid=self.zone.id,
            templateid=self.template.id,
            accountid=self.account.name,
            domainid=self.account.domainid,
            serviceofferingid=self.service_offering.id,
            overridediskofferingid=disk_off_id,
            diskofferingid=disk_off_id,
            size=2,
            hypervisor=self.hypervisor,
            rootdisksize=10
        )
        self.cleanup.append(virtual_machine_tier2_template)
```

File: test/integration/plugins/storpool/TestStorPoolTiers.py [hunk 12]
```diff
@@ -0,0 +1,554 @@
...
+            templateid=self.template.id,
+            accountid=self.account.name,
+            domainid=self.account.domainid,
+            serviceofferingid=self.service_offering.id,
+            overridediskofferingid=self.disk_offerings_tier1_tags.id,
+            diskofferingid=self.disk_offerings_tier1_tags.id,
+            size=2,
+            hypervisor=self.hypervisor,
+            rootdisksize=10
+        )
```
Comment (DaanHoogland) [inline test/integration/plugins/storpool/TestStorPoolTiers.py:464]:
```suggestion
        virtual_machine_tier1_tag = VirtualMachine.create(
            self.apiclient,
            {"name": "StorPool-%s" % uuid.uuid4()},
            zoneid=self.zone.id,
            templateid=self.template.id,
            accountid=self.account.name,
            domainid=self.account.domainid,
            serviceofferingid=self.service_offering.id,
            overridediskofferingid=self.disk_offerings_tier1_tags.id,
            diskofferingid=self.disk_offerings_tier1_tags.id,
            size=2,
            hypervisor=self.hypervisor,
            rootdisksize=10
        )
        self.cleanup.append(virtual_machine_tier1_tag)
```

File: test/integration/plugins/storpool/TestStorPoolTiers.py (hunk 7 above)
Comment (slavkap) [reply test/integration/plugins/storpool/TestStorPoolTiers.py:254]: Thanks @DaanHoogland, for the review! I'm adding the account to be cleaned and when the cleanup is invoked it destroys everything on that account (VMs, volumes, snapshots, etc.)

```
        cls.account = cls.helper.create_account(
            cls.apiclient,
            cls.services["account"],
            accounttype=1,
            domainid=cls.domain.id,
            roleid=1
        )
        cls._cleanup.append(cls.account)
```

Comment (DaanHoogland) [reply test/integration/plugins/storpool/TestStorPoolTiers.py:254]: ok, no worries or :-1: just remarks

File: test/integration/plugins/storpool/TestStorPoolTiers.py
Comment (slavkap) [reply test/integration/plugins/storpool/TestStorPoolTiers.py:1]: thanks @DaanHoogland, I've renamed the file

File: test/integration/plugins/storpool/TestStorPoolTiers.py (hunk 4 above)
Comment (slavkap) [reply test/integration/plugins/storpool/TestStorPoolTiers.py]: applied in the latest commit

File: test/integration/plugins/storpool/TestStorPoolTiers.py (hunk 6 above)
Comment (slavkap) [reply test/integration/plugins/storpool/TestStorPoolTiers.py]: applied in the latest commit

File: test/integration/plugins/storpool/TestStorPoolTiers.py (hunk 5 above)
Comment (slavkap) [reply test/integration/plugins/storpool/TestStorPoolTiers.py]: as I mentioned in one of the suggestions, I don't add all objects to be cleaned. Everything created under the account created from the tests is destroyed at the end

File: test/integration/plugins/storpool/test-storpool-tiers.py [hunk 13]
```diff
@@ -0,0 +1,549 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (DaanHoogland) [inline test/integration/plugins/storpool/test-storpool-tiers.py:1]: @slavkap , file name convention dictates '_'s and not '-'s . can you rename this file?

File: test/integration/plugins/storpool/test-storpool-tiers.py [hunk 14]
```diff
@@ -0,0 +1,549 @@
...
+            cls.resourceDetails(qos, disk_offerings.id, )
+        return disk_offerings
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.cleanUpCloudStack()
+
+    @classmethod
+    def cleanUpCloudStack(cls):
+        super(TestStorPoolTiers, cls).tearDownClass()
```
Comment (DaanHoogland) [inline test/integration/plugins/storpool/test-storpool-tiers.py:190]: why the double redirection and not just call super() directly?

File: test/integration/plugins/storpool/test-storpool-tiers.py (hunk 13 above)
Comment (slavkap) [reply test/integration/plugins/storpool/test-storpool-tiers.py]: done

File: test/integration/plugins/storpool/test-storpool-tiers.py (hunk 14 above)
Comment (slavkap) [reply test/integration/plugins/storpool/test-storpool-tiers.py]: thanks, I've  removed the unnecessary method

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by DaanHoogland** [comment]
not sure if this requires more testing than smoke tests (for other storage solutions cc @harikrishna-patnala @rp- @rg9975 ) ???

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rp-** [comment]
> not sure if this requires more testing than smoke tests (for other storage solutions cc @harikrishna-patnala @rp- @rg9975 ) ???

I don't think this can in anyway break the Linstor driver

**Review by rg9975 — APPROVED** [review]

**Pull request merged**